### PR TITLE
Comments: Add support for custom emojis + member info

### PR DESCRIPTION
**Pull request by ChunkyProgrammer**
closes https://github.com/iv-org/invidious/issues/2308

Test video (pinned comment has custom emojis):
https://www.youtube.com/watch?v=v3wm83zoSSY

Test video (pinned comment has emojis, not custom)
https://www.youtube.com/watch?v=Dp2eu85uOf0

Invidious screenshot:
Before:
image

After:
image

YouTube screenshots:
Web:
image

Mobile:
image

### CODE REVIEW

**Comment by ChunkyProgrammer** [comment]
I don't want to be awarded the bounty

**Comment by unixfox** [comment]
funny

File: src/invidious/comments.cr [hunk 1]
```diff
@@ -674,6 +689,18 @@ def content_to_comment_html(content, video_id : String? = "")
     text = "<b>#{text}</b>" if run["bold"]?
     text = "<s>#{text}</s>" if run["strikethrough"]?
     text = "<i>#{text}</i>" if run["italics"]?
+    if run["emoji"]?
```
Comment (SamantazFox) [inline src/invidious/comments.cr:692]: can you add a new line before that if block, for clarity?

File: src/invidious/comments.cr [hunk 2]
```diff
@@ -674,6 +689,18 @@ def content_to_comment_html(content, video_id : String? = "")
...
+      if run["emoji"]["isCustomEmoji"]?.try &.as_bool
+        if emojiImage = run.dig?("emoji", "image")
+          emojiAlt = emojiImage.dig?("accessibility", "accessibilityData", "label").try &.as_s || text
+          emojiThumb = emojiImage["thumbnails"][0]
+          text = "<img alt=\"#{emojiAlt}\" src=\"/ggpht#{URI.parse(emojiThumb["url"].as_s).request_target}\" title=\"#{emojiAlt}\" width=\"#{emojiThumb["width"]}\" height=\"#{emojiThumb["height"]}\" style=\"margin-right:2px;margin-left:2px;\"/>"
```
Comment (SamantazFox) [inline src/invidious/comments.cr:697]: For clarity, please use either `String.build` or [Heredocs] (`<<-HTML`).
If you're using `String.build`, you can use [percent string literals](https://crystal-lang.org/reference/1.6/syntax_and_semantics/literals/string.html#percent-string-literals) `%()` to write strings with a lot of quotes. E.g:
```crystal
output_string = String.build do |str|
  str << %(<img alt=") << emojiAlt << '"'
  # etc...
  str << %( style="margin-right:2px;margin-left:2px;"/>)
end
```

**Review by SamantazFox — CHANGES_REQUESTED** [review]

File: src/invidious/comments.cr [hunk 3]
```diff
@@ -182,7 +182,11 @@ def fetch_youtube_comments(id, cursor, format, locale, thin_mode, region, sort_b
               json.field "contentHtml", content_html
 
               json.field "isPinned", (node_comment["pinnedCommentBadge"]? != nil)
-
+              json.field "isMember", (node_comment["sponsorCommentBadge"]? != nil)
```
[the remark this answers is not in the document]
Comment (ChunkyProgrammer) [reply src/invidious/comments.cr]: Sounds good, I updated the title text as well

File: src/invidious/comments.cr (hunk 2 above)
Comment (ChunkyProgrammer) [reply src/invidious/comments.cr]: Done 🙂

File: src/invidious/comments.cr [hunk 4]
```diff
@@ -675,6 +689,27 @@ def content_to_comment_html(content, video_id : String? = "")
...
+      if run["emoji"]["isCustomEmoji"]?.try &.as_bool
+        if emojiImage = run.dig?("emoji", "image")
+          emojiAlt = emojiImage.dig?("accessibility", "accessibilityData", "label").try &.as_s || text
+          emojiThumb = emojiImage["thumbnails"][0]
+          text = String.build do |str|
+            str << %(<img alt=") << emojiAlt << '"'
+            str << %(src="/ggpht) << URI.parse(emojiThumb["url"].as_s).request_target << '"'
+            str << %(title=") << emojiAlt << '"'
+            str << %(width=") << emojiThumb["width"] << '"'
+            str << %(height=") << emojiThumb["height"] << '"'
```
Comment (SamantazFox) [inline src/invidious/comments.cr:703]: watch out there, your HTML attributes aren't separated by a space!

Comment (ChunkyProgrammer) [reply src/invidious/comments.cr]: Fixed 🙂

File: src/invidious/comments.cr [hunk 5]
```diff
@@ -675,6 +689,27 @@ def content_to_comment_html(content, video_id : String? = "")
...
+        if emojiImage = run.dig?("emoji", "image")
+          emojiAlt = emojiImage.dig?("accessibility", "accessibilityData", "label").try &.as_s || text
+          emojiThumb = emojiImage["thumbnails"][0]
+          text = String.build do |str|
+            str << %(<img alt=") << emojiAlt << "\" "
+            str << %(src="/ggpht) << URI.parse(emojiThumb["url"].as_s).request_target << "\" "
+            str << %(title=") << emojiAlt << "\" "
+            str << %(width=") << emojiThumb["width"] << "\" "
+            str << %(height=") << emojiThumb["height"] << "\" "
+            str << %(style="margin-right:2px;margin-left:2px;"/>)
```
Comment (SamantazFox) [inline src/invidious/comments.cr:704]: Can you create a dedicated class (e.g `channel-emoji`) ad put the styles in `default.css`?
I'd like to see all inline CSS to be removed in the future, so it's better if we can start now ^^

**Review by SamantazFox — APPROVED** [review]

**Comment by SamantazFox** [comment]
I'm getting the following error on the test instance (`3636.diff` freshly applied on top of `master`):


Title: `Missing hash key: "isSponsor" (KeyError)`
Date: `2023-03-07T20:49:24Z`
Route: `/channel/UCRELfrtyz_teOEQXjF04BIA/community`
Version: `2023.03.07-d79d6f38 @ master`

<details>
<summary>Backtrace</summary>
<p>
   
```
Missing hash key: "isSponsor" (KeyError)
  from /usr/share/crystal/src/hash.cr:1077:11 in '[]'
  from /usr/share/crystal/src/json/any.cr:103:7 in '[]'
  from invidious.git/src/invidious/comments.cr:337:10 in 'template_youtube_comments'
  from invidious.git/src/invidious/comments.cr:300:1 in 'template_youtube_comments'
  from src/invidious/views/community.ecr:30:13 in 'community'
  from invidious.git/src/invidious/routing.cr:123:37 in '->'
  from invidious.git/lib/kemal/src/kemal/route.cr:12:26 in '->'
  from invidious.git/src/invidious/helpers/handlers.cr:30:37 in 'process_request'
  from invidious.git/lib/kemal/src/kemal/route_handler.cr:17:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/lib/kemal/src/kemal/websocket_handler.cr:13:14 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/lib/kemal/src/kemal/filter_handler.cr:21:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/src/invidious/helpers/handlers.cr:212:5 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/src/invidious/helpers/handlers.cr:94:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/src/invidious/helpers/handlers.cr:145:12 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/src/invidious/helpers/handlers.cr:70:5 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/src/ext/kemal_static_file_handler.cr:162:16 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/lib/kemal/src/kemal/exception_handler.cr:8:7 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/src/invidious/helpers/logger.cr:17:35 in 'call'
  from /usr/share/crystal/src/http/server/handler.cr:28:7 in 'call_next'
  from invidious.git/lib/kemal/src/kemal/init_handler.cr:12:7 in 'call'
  from /usr/share/crystal/src/http/server/request_processor.cr:51:11 in 'process'
  from /usr/share/crystal/src/http/server.cr:515:5 in 'handle_client'
  from /usr/share/crystal/src/http/server.cr:468:13 in '->'
  from /usr/share/crystal/src/fiber.cr:146:11 in 'run'
  from /usr/share/crystal/src/fiber.cr:98:34 in '->'
  from ???
```
</p>
</details>

**Comment by ChunkyProgrammer** [comment]
It should be fixed now 🙂

**Comment by SamantazFox** [comment]
> Ah, i didn't think of that. I guess [#3636 (comment)](https://github.com/iv-org/invidious/pull/3636#discussion_r1110211995) was needed for community posts (i had initially just copied the way that verified does it but wasnt sure why)

Duuuh, I misread the code's context again, sorry :/

**Review by SamantazFox — APPROVED** [review]

**Comment by SamantazFox** [comment]
Tested on https://test.invidious.io, works perfectly!

**Comment by SamantazFox** [comment]
Thanks for contributing to Invidious ^^

**Pull request merged**